lib/streamlit/caching.py

```diff
@@ -115,7 +92,19 @@ def _show_cached_st_function_warning(dg):
     # Avoid infinite recursion by suppressing additional cached
     # function warnings from within the cached function warning.
     with suppress_cached_st_function_warning():
-        dg.warning(CACHED_ST_FUNCTION_WARNING)
+        e = CachedStFunctionWarning(
+            """
+Your script writes to your Streamlit app from within a cached function. This
+code will only be called when we detect a cache "miss", which can lead to
+unexpected results.
+
+How to resolve this warning:
+* Move the Streamlit function call outside the cached function.
+* Or, if you know what you're doing, use `@st.cache(suppress_st_warning=True)`
+to suppress the warning.
+"""
+        )
+        dg.exception(e)
```

Comment:
We now use Exception objects for cache-related warnings, so their text is exactly like the Exception text (rather than trying to emulate it by hand).